effectful/handlers/pyro.py

```diff
@@ -252,190 +264,280 @@ def __repr__(self):
         return f"Naming({self.name_to_dim})"
 
 
-class PositionalDistribution(pyro.distributions.torch_distribution.TorchDistribution):
-    """A distribution wrapper that lazily converts indexed dimensions to
-    positional.
+@defop
+def named_distribution(
+    d: Annotated[TorchDistribution, Scoped[A | B]],
+    *names: Annotated[Operation[[], int], Scoped[B]],
+) -> Annotated[TorchDistribution, Scoped[A | B]]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            if dist_constr is dist.Independent:
+                base_dist, reinterpreted_batch_ndims = args
+                return dist.Independent(
+                    named_distribution(
+                        typing.cast(TorchDistribution, base_dist), *names
+                    ),
+                    reinterpreted_batch_ndims,
+                )
+            else:
+                named_args = []
+                for a in args:
+                    assert isinstance(a, torch.Tensor)
+                    named_args.append(
+                        Indexable(typing.cast(torch.Tensor, a))[
+                            tuple(n() for n in names)
+                        ]
+                    )
+                assert callable(dist_constr)
+                return dist_constr(*named_args)
+        case _:
+            raise NotImplementedError
 
-    """
 
-    indices: Mapping[Operation[[], int], int]
+@defop
+def positional_distribution(
+    d: Annotated[TorchDistribution, Scoped[A]],
+) -> Tuple[TorchDistribution, Naming]:
+    match defterm(d):
+        case Term(op=_call, args=(dist_constr, *args)) if _call is call:
+            if dist_constr is dist.Independent:
```

Comment:
As in `named_distribution`, I don't think this special case handling of `Independent` is necessary. You just need to distribute `positional_distribution` across the arguments - something like this:
```py
d = defterm(d) if not isinstance(d, Term) else d
if not (d.op is call and issubclass(d.args[0], TorchDistribution)):
  raise NotImplementedError

dist_constr, args = args[0], args[1:]
naming = ...
indices = ...

def _to_positional(a):
  if isinstance(a, torch.Tensor):
    return to_tensor(a, indices)
  elif isinstance(a, TorchDistribution):
    return positional_distribution(a)[0]
  else:
    return a
 
return dist_constr(*tree.map_structure(_to_positional, args)), naming
```